### PR TITLE
[Enhancement] Add hashJoin execution cost model

### PR DESCRIPTION
**Pull request by packy92**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

For different distributed join execution methods, due to their different execution characteristics, they should have different cost. We currently have four execution modes: broadcast join, bucket_shuffle join, shuffle join, and colocate join. The latter three are essentially shuffle joins, so their execution efficiency can be approximately regarded as consistent. The implementation of broadcast join is a little different from the other three.
1. In the build hash table stage of broadcast join, the parallelism is 1 while it can be parallized in shuffle join.
2. Broadcast join consumes more memory for redundant data in multiple BEs.
3. The data size in each hash table of shuffle join is 1/parallelism of right table_size while it is full size in broadcast join. A small hash table can improve probe efficiency.
4. The process of redistributing data by shuffle operation makes the tasks of each processing thread are more balanced, and it is more conducive to parallel computing.
Therefore, our join execution cost calculation model should comprehensively consider parameters such as join execution mode, parallelism, and left and right table size. The most important thing in the model is the evaluation of the probe cost for each row. When the size of the right table is greater than bottom_number, the average probe cost needs to be expanded. The parallel computing characteristics of shuffle join can offset part of the probe cost. We also set an upper limit on the probe cost to avoid cost distortion caused by too large table data.

## Performance Test
Environment benchmark07.
TPCDS-1t from 806s to 637s.
TPCH-1t from 254s to 250s.

## UT changes
 - hive sql plan chanages. New join execution cost model inclined to shuffle join when the right table is larger than a million level. Cases with a 10M level table join with a 1M level table use shuffle join now. The changes of q2.sql reflects a bug. When the child's best plan was updated, his father's confidence statistics were not updated because of oldChildcost + oldEnforcerCost < newChildCost(smaller than old) + newEnforcerCost(bigger than old).
  
 - enumerate paln numbers changes. Join cost number has different from before which may lead to different prune operations.
 - Sqls in testCountPruneJoinByProject, testJoinWithPipelineDop use broadcast join before is very slow.  This pr optimizes them.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        for (BinaryPredicateOperator predicateOperator : eqOnPredicates) {
+            ColumnRefOperator leftCol = (ColumnRefOperator) predicateOperator.getChild(0);
+            ColumnRefOperator rightCol = (ColumnRefOperator) predicateOperator.getChild(1);
+            if (context.getChildStatistics(1).getColumnStatistics().containsKey(leftCol)) {
+                keySize += context.getChildStatistics(1).getColumnStatistic(leftCol).getAverageRowSize();
+            } else if (context.getChildStatistics(1).getColumnStatistics().containsKey(rightCol)) {
+                keySize += context.getChildStatistics(1).getColumnStatistic(rightCol).getAverageRowSize();
+            }
+        }
+        double degradeRatio;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java:119]: what's the degrade meaning here?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            degradeRatio = Math.max(1, (Math.log(mapSize / BOTTOM_NUMBER) -
+                    Math.log(parallelFactor) / Math.log(2)));
+            // normalize ration when it hits the limit
+            degradeRatio = Math.min(SHUFFLE_MAX_RATIO, degradeRatio);
+        }
+        LOG.debug("execMode: {}, degradeRatio: {}", execMode, degradeRatio);
+        return degradeRatio;
+    }
+
+    private String deriveJoinExecMode() {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java:138]: is enum better than string?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 3]
```diff
@@ -35,26 +37,54 @@
...
+
+    private static final Logger LOG = LogManager.getLogger(CostModel.class);
     public static double calculateCost(GroupExpression expression) {
         ExpressionContext expressionContext = new ExpressionContext(expression);
         return calculateCost(expressionContext);
     }
 
     private static double calculateCost(ExpressionContext expressionContext) {
-        CostEstimator costEstimator = new CostEstimator();
+        CostEstimator costEstimator = new CostEstimator(null);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:55]: prefer empty list than null

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java:119]: When right table size grows, the probe efficiency will degrade. Use the degradeRation to distinguish the costs for different join situation.

Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java:114]: I see. but I think maybe we could have a clear name, such as memoryCachePenaltyRatio ?

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/HashJoinCostModel.java]: Sounds good.

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 99 / 102 (97.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalDistributionOperator.java|0|1|00.00%|[54]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostEstimate.java|1|2|50.00%|[72]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|35|36|97.22%|[234]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/HashJoinCostModel.java|60|60|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|3|100.00%|[]|

**Review by kangkaisen — DISMISSED** [review]

**Comment by packy92** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 99 / 102 (97.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalDistributionOperator.java|0|1|00.00%|[54]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostEstimate.java|1|2|50.00%|[72]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|35|36|97.22%|[234]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/HashJoinCostModel.java|60|60|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|3|100.00%|[]|

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/EnumeratePlanTest.java [hunk 4]
```diff
@@ -48,7 +48,7 @@ public void testTPCHQ4EnumPlan() {
 
     @Test
     public void testTPCHQ5EnumPlan() {
-        runFileUnitTest("enumerate-plan/tpch-q5");
+        runFileUnitTest("enumerate-plan/tpch-q5", true);
```
Comment (Seaven) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/EnumeratePlanTest.java:51]:
```suggestion
        runFileUnitTest("enumerate-plan/tpch-q5");
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 99 / 102 (97.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalDistributionOperator.java|0|1|00.00%|[54]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostEstimate.java|1|2|50.00%|[72]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|35|36|97.22%|[234]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/HashJoinCostModel.java|60|60|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|3|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by packy92** [comment]
starrocks_admit_test

**Comment by packy92** [comment]
run starrocks_admit_test

**Pull request merged**